filterByTag.js

```diff
@@ -0,0 +1,82 @@
+class Filter {
+  constructor(tagSelector) {
+    this.tagSelector = document.querySelectorAll(tagSelector); // What tag to filter
+    this.activeTags = []; //Array to store active tags
+  }
+
+  /*  Filter method to display the selected tags and hide the other tags.
+        This will also reset the filter and display all challenges when none is selected */
+  filterTags() {
+    const cards = document.querySelectorAll('.challenges-container__challenge');
+
+    if (this.activeTags.length === 0) {
+      cards.forEach((card) => {
+        card.style.display = '';
+      });
+    } else {
+      cards.forEach((card) => {
+        if (this.activeTags.some((tag) => card.classList.contains(tag))) {
+          card.style.display = '';
+        } else {
+          card.style.display = 'none';
+        }
+      });
+    }
+  }
+
+  // Method to Toggle the chosen tag as active
+  toggleTag(tag) {
+    const index = this.activeTags.indexOf(tag);
+    if (index === -1) {
+      this.activeTags.push(tag);
+    } else {
+      this.activeTags.splice(index, 1);
+    }
+  }
+}
+
+// class for eventlistener and handlers for the tag buttons
+class FilterButton {
+  constructor(buttonId, filterInstance, tag) {
+    this.button = document.getElementById(buttonId);
+    this.filterInstance = filterInstance;
+    this.tag = tag;
+
+    this.button.addEventListener('click', this.handleButtonClick.bind(this));
+  }
+
+  handleButtonClick() {
+    this.filterInstance.toggleTag(this.tag);
+    this.filterInstance.filterTags();
+
+    //Add class -active to the active tag button
+    if (this.filterInstance.activeTags.includes(this.tag)) {
```

Comment:
Now I'm being picky, but only because I can tell that you're trying to learn object-oriented programming and I want to give you as much information as possible.

What I'm seeing here is a "Law of Demeter" violation, or a "code smell" that is usually called "message chains". Basically what it means is that this class `FilterButton` is not just interacting with `Filter` (`this.filterInstance`) but also with it's `activeTags` property's method `includes`.

This means that if at some point you want to restructure `Filter` so that it doesn't use an array anymore for `activeTags`, you also have to change this code in a separate class. This is proof of poor separation of concerns (because if separation was good, change in one place should not require change in another place).

A good way of solving this would be to add a method on `Filter` called `hasTag()` that internally does `this.activeTags.includes(tag)`, e.g.:

```
// This goes in Filter
hasTag(tag) {
  return this.activeTags.includes(tag);
}
```

That way, you could do `this.filterInstance.hasTag(this.tag)` here.

With that change, `Filter` is no longer exposing it's `activeTags` array, and if you wanted to change that to an object, Map or some other collection, you could just make those changes in `Filter`, update it's own `hasTag()` method and this code here would just keep working.